contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,207 @@
+
+    // Used for sending and receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 sourceBlockchainID_,
+        address nativeTokenSourceAddress_,
+        uint256 initialReserveImbalance_
+    ) {
+        currentBlockchainID = WarpMessenger(
+            0x0200000000000000000000000000000000000005
+        ).getBlockchainID();
+
+        require(
+            teleporterMessengerAddress != address(0),
+            "NativeTokenDestination: zero TeleporterMessenger Address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(
+            sourceBlockchainID_ != bytes32(0),
+            "NativeTokenDestination: zero Source Chain ID"
+        );
+        require(
+            sourceBlockchainID_ != currentBlockchainID,
+            "NativeTokenDestination: Cannot Bridge With Same Blockchain"
+        );
+        sourceBlockchainID = sourceBlockchainID_;
+
+        require(
+            nativeTokenSourceAddress_ != address(0),
+            "NativeTokenDestination: zero Source Contract Address"
+        );
+        nativeTokenSourceAddress = nativeTokenSourceAddress_;
+
+        initialReserveImbalance = initialReserveImbalance_;
+        currentReserveImbalance = initialReserveImbalance_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "NativeTokenDestination: Unauthorized TeleporterMessenger contract"
+        );
+
+        // Only allow messages from the source chain.
+        require(
+            senderBlockchainID == sourceBlockchainID,
+            "NativeTokenDestination: Invalid Source Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(
+            senderAddress == nativeTokenSourceAddress,
+            "NativeTokenDestination: Unauthorized Sender"
+        );
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(
+            recipient != address(0),
+            "NativeTokenDestination: zero Recipient Address"
+        );
+        require(amount != 0, "NativeTokenDestination: Transfer value of 0");
+
+        uint256 adjustedAmount = amount;
+        if (currentReserveImbalance > 0) {
+            if (amount > currentReserveImbalance) {
+                emit CollateralAdded({amount: currentReserveImbalance, remaining: 0});
```

Comment:
nit: let's emit the event after the state change

Reply:
This one is this way to not have to have an additional placeholder variable or do an extra calculation. I can still switch it if you want

Reply:
Ah, gotcha. I'm fine leaving as is.